back/api/v1/authentication.go

```diff
@@ -0,0 +1,168 @@
+package v1
+
+import (
+	"4ctf/models"
+	"4ctf/utils"
+	"4ctf/views"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/savsgio/atreugo/v11"
+	"github.com/volatiletech/sqlboiler/v4/boil"
+)
+
+func setupAuthRoutes(api *Api, router *atreugo.Router) {
+	group := router.NewGroupPath("/auth")
+
+	group.POST("/login", WithDefaults(api, WithBody(login)))
+	group.POST("/register", WithDefaults(api, WithBody(register)))
+	group.GET("/logout", WithDefaults(api, logout))
+	group.GET("/me", WithDefaults(api, WithValidSession(profile)))
+}
+
+type LoginRequest struct {
+	Username string `json:"username" validate:"min=2,max=40,regexp=^[a-zA-Z]*$"`
+	Password string `json:"password" validate:"max=40,password"`
+}
+
+func login(api *Api) func(ctx *atreugo.RequestCtx, lr *LoginRequest) error {
+	return func(ctx *atreugo.RequestCtx, lr *LoginRequest) error {
+		// Check if the user exists in the database
+		user, err := models.Users(models.UserWhere.Username.EQ(lr.Username)).OneG(ctx)
+		if err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				api.
+					WithField("username", lr.Username).
+					WithError(err).
+					Error("cannot find user")
+			} else {
+				api.
+					WithField("username", lr.Username).
+					WithError(err).
+					Warn("cannot find user")
+			}
+			return ctx.JSONResponse(NewErrorResponse(404, []ResponseError{{Message: "User not found"}}))
+		}
+
+		// Check if the password is correct
+		if !utils.VerifyPassword(lr.Password, user.PasswordHash) {
+			api.
+				WithField("username", lr.Username).
+				Warn("invalid password")
+			return ctx.JSONResponse(NewErrorResponse(404, []ResponseError{{Message: "User not found"}}))
+		}
+
+		// Create a new userSession
+		userSession := models.UserSession{
+			UserID:    user.ID,
+			ExpiresAt: time.Now().Add(30 * 24 * time.Hour),
+		}
+
+		if err := userSession.InsertG(ctx, boil.Whitelist(models.UserSessionColumns.UserID, models.UserSessionColumns.ExpiresAt)); err != nil {
```

Comment:
In which case could we have an error on the session's creation ?

Reply:
Many cases : 
- Database disconnects
- Connection close at the wrong time
- Unknown error on the database side

Golang is made to return error whenever something bad can happen : 
> Anything that can go wrong, will go wrong

Reply:
ok thx for explanations